ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoMaterialProperties.cpp

```diff
@@ -0,0 +1,294 @@
+
+    {  // Local Newton solver
+        using LocalResidualVector = Eigen::Matrix<double, 2, 1>;
+        using LocalUnknownVector = Eigen::Matrix<double, 2, 1>;
+        LocalJacobianMatrix J_loc;
+
+        auto const update_residual = [&](LocalResidualVector& residual) {
+            calculateResidual(PG, X, Sw, X_m, residual);
+        };
+
+        auto const update_jacobian = [&](LocalJacobianMatrix& jacobian) {
+            calculateJacobian(t, x, PG, X, jacobian, Sw,
+                              X_m);  // for solution dependent Jacobians
+        };
+
+        auto const update_solution = [&](LocalUnknownVector const& increment) {
+            // increment solution vectors
+            Sw += increment[0];
+            X_m += increment[1];
+        };
+
+        // TODO Make the following choice of maximum iterations and convergence
+        // criteria available from the input file configuration:
+        const int maximum_iterations(20);
+        const double tolerance(1.e-14);
+
+        auto newton_solver = NumLib::NewtonRaphson<
+            decltype(linear_solver), LocalJacobianMatrix,
+            decltype(update_jacobian), LocalResidualVector,
+            decltype(update_residual), decltype(update_solution)>(
+            linear_solver, update_jacobian, update_residual, update_solution,
+            maximum_iterations, tolerance);
+
+        auto const success_iterations = newton_solver.solve(J_loc);
+
+        if (!success_iterations)
+            return false;
+
+        // If the Newton loop didn't run, the linear solver will not be
+        // initialized.
+        // This happens usually for the first iteration of the first timestep.
+        if (*success_iterations == 0)
+            linear_solver.compute(J_loc);
+    }
+    dsw_dpg = Calculate_dSwdP(PG, Sw, X_m);
+    dsw_dX = Calculate_dSwdX(PG, X, Sw, X_m);
+    dxm_dpg = Calculate_dX_mdP(PG, Sw, X_m, dsw_dpg);
+    dxm_dX = Calculate_dX_mdX(PG, Sw, X_m, dsw_dX);
+    return true;
+}
+void TwoPhaseFlowWithPrhoMaterialProperties::calculateResidual(
+    double const PL, double const X, double Sw, double rho_h2_wet,
+    ResidualVector& res)
+{
+    const double PG =
+        PL +
+        _capillary_pressure_models[_current_material_id]->getCapillaryPressure(
+            Sw);
+    const double rho_h2_nonwet = PG * molar_mass_h2 / R / 303.15;
+
+    // calculating residual
+    res(0) = Calc_equili_rho_m(PG, Sw, rho_h2_wet);
+    res(1) = Calc_Saturation(PL, X, Sw, rho_h2_wet, rho_h2_nonwet, 303.15);
+}
+
+void TwoPhaseFlowWithPrhoMaterialProperties::calculateJacobian(
+    double const /*t*/, ProcessLib::SpatialPosition const& /*x*/, double const PL,
+    double const /*X*/, JacobianMatrix& Jac, double Sw, double rho_h2_wet)
+{
+    const double PG =
+        PL +
+        _capillary_pressure_models[_current_material_id]->getCapillaryPressure(
+            Sw);
+    const double rho_h2_nonwet = PG * molar_mass_h2 / R / 303.15;
```

Comment:
This number 303.15 is popping up in multiple places. Please add comment what it is (units, usage, etc.).
It would be better if there would be a single place for magic numbers.